### PR TITLE
ISPN-11297 Prevent creating clustered caches incompatible with global state

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-11297

### CODE REVIEW

File: core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java [hunk 1]
```diff
@@ -85,18 +91,42 @@ void start() {
       parserRegistry = new ParserRegistry();
 
       localConfigurationManager.initialize(cacheManager, configurationManager, executorService);
+
+      // Load any state we previously had
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java]: This is kind of confusing, both the local persistent state and the state cache can be said to hold existing configuration state. Could you rename the variables a bit to make it clear where each came from?

File: core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java [hunk 2]
```diff
@@ -172,11 +202,15 @@ void start() {
...
       log.debugf("Create cache %s", name);
-      ConfigurationBuilderHolder builderHolder = parserRegistry.parse(state.getConfiguration());
-      Configuration configuration = builderHolder.getNamedConfigurationBuilders().get(name).build();
+      Configuration configuration = buildConfiguration(name, state);
       return localConfigurationManager.createCache(name, state.getTemplate(), configuration, state.getFlags());
    }
 
+   private Configuration buildConfiguration(String name, CacheState state) {
+      ConfigurationBuilderHolder builderHolder = parserRegistry.parse(state.getConfiguration());
+      return builderHolder.getNamedConfigurationBuilders().get(name).build();
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java]: I'm not sure if this is ok. Some attributes are validated against the global configuration, and they may not be valid with the current global configuration.

File: core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java [hunk 3]
```diff
@@ -49,12 +49,18 @@
...
    public static final String CACHE_SCOPE = "cache";
 
-   @Inject EmbeddedCacheManager cacheManager;
-   @Inject LocalTopologyManager localTopologyManager;
-   @Inject ConfigurationManager configurationManager;
-   @Inject InternalCacheRegistry internalCacheRegistry;
-   @Inject GlobalComponentRegistry globalComponentRegistry;
-   @Inject @ComponentName(KnownComponentNames.ASYNC_OPERATIONS_EXECUTOR)
+   @Inject
+   EmbeddedCacheManager cacheManager;
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java]: That was my IDE reformatting at work :)

File: core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java (hunk 2 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java]: I now build against the global configuration

File: core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java (hunk 1 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java]: I have renamed things to make things clearer

**Comment by danberindei** [comment]
Merged, thanks Tristan1

**Pull request merged**